### DISCUSSION TITLE
vim syntax highlighting

### DISCUSSION THREAD

**Discussion by georgalis** [Homebrew/discussions#4247] [Tap maintenance and Homebrew development] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.3
ORIGIN: https://github.com/Homebrew/brew
HEAD: ba6d87eed9ce31f42fa2b33a4e6067d9cf17f13f
Last commit: 6 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: fe748547bbe691f4c5693e594b0dd9d87993f3e5
Core tap last commit: 2 days ago
Core tap branch: master
Core tap JSON: 20 Feb 23:10 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: /opt/homebrew/bin/vim
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.39.2 => /opt/homebrew/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2.1-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

My recently updated brew vim has broken syntax highlighting for bash...
../Cellar/vim/9.0.1300/bin/vim

this is most apparent with the use of line continuation '\'  AND and OR control operator. eg
<pre>
[ "a" = "b" ] \
    && echo true \
    || echo false
</pre>

I note the OSX vim is working fine
/usr/bin/vim --version
VIM - Vi IMproved 9.0 (2022 Jun 28, compiled Dec 16 2022 23:29:16)

What is the most expedient way to raise this with appropriate developers?

How do I determine the prior version(s) of vim and configure brew to use that?

Thanks!

**Comment by georgalis** [comment]
This issue was opened and closed in github vim... the fix is
Downgrade sh.vim syntax from version 205 to 204.
Although I don't see the fix applied to the HEAD (release) yet.
https://github.com/vim/vim/issues/11937#issuecomment-1421116531

My first workaround is simply to use `/usr/bin/vim` however
I would very much like to understand how to configure
brew to use the last working formula that was installed.

I used `brew log vim` and found:
<pre>
commit d24021129dba6149a352cb742bd331da5288c22f
Author: BrewTestBot <1589480+BrewTestBot@users.noreply.github.com>
Date:   Fri Jan 27 22:04:51 2023 +0000

    vim: update 9.0.1250 bottle.
</pre>

Through some github black magic, I was able to locate the corresponding raw package formula
https://raw.githubusercontent.com/Homebrew/homebrew-core/d24021129dba6149a352cb742bd331da5288c22f/Formula/vim.rb

I tried to invoke brew install with this to rollback to the last working vim (syntax highlighting), but no command joy. This operation seems so simple and essential, am I doing something wrong? Finding the working ruby file, this file, and stalling on the next step of getting brew to revert what broke with my last upgrade...  clearly I'm not understanding the brew UX model?

 Here;s my workaround... 
<pre>
mv /opt/homebrew//Cellar/vim/9.0.1300/share/vim/vim90/syntax/sh.vim /opt/homebrew//Cellar/vim/9.0.1300/share/vim/vim90/syntax/sh.vim~
wget https://raw.githubusercontent.com/vim/vim/86b4816766d976a7ecd4403eca1f8bf6b4105800/runtime/syntax/sh.vim
mv sh.vim /opt/homebrew//Cellar/vim/9.0.1300/share/vim/vim90/syntax/sh.vim
</pre>

**Reply by georgalis** [reply]
...
`mv sh.vim /opt/homebrew//Cellar/vim/9.0.1350/share/vim/vim90/syntax/sh.vim
`

**Reply by dsnyder0pc** [reply]
Thanks so much, @georgalis . That works for me. Hoping the vim team will fix this properly in HEAD soon. I've not read all of the gory details, but leaving this broken for so many _innocent_ users is not good form.

**Reply by bat9r** [reply]
Ohh, my 
I was searching for this answer pretty long
Many thanks!

**Comment by georgalis** [comment, marked as answer]
@dsnyder0pc I'm glad that worked for you. It's my second issue with brew, and they were both closed without resolution or even addressing my carefully described concerns. The problem with the brew "evergreen" philosophy comes when it installs software that does not share the same evergreen philosophy. As the case of vim, the software authors expect users to (have the ability to) choose the version they want, not simply follow HEAD. I have had great success with pkgsrc in hpc platforms. The quarterly release process has matured for 77 quarters now, almost 20 years of focus on uptime, security, and reliability, in the face of bugs and complex dependency resolution, for server infrastructure---it addresses my concerns. This issue has led me to reevaluate the mac binary support, and I'm happy to report a painless (:15 minute) transition to pkgin package binaries on mac, which also enables me to use the same release software and versions, across all my platforms: mac, linux, or bsd in the datacenter. 
https://github.com/Homebrew/brew/issues/6637
https://github.com/Homebrew/brew/issues/14809
Binary packages: https://pkgsrc.smartos.org/install-on-macos/
Source packages: https://pkgsrc.org/

**Reply by alebcay** [reply]
Regarding https://github.com/Homebrew/brew/issues/14809, I responded carefully regarding your request to be able to identify older working versions, roll back, and pin. The gist of my response is that there is no way to seamlessly support this without essentially rewriting a significant portion of `brew`. I did also provide a workaround for this "evergreen" nature using the `brew extract` command.

I understand this is probably not the resolution that you hoped for ("probably won't fix, at least anytime soon" vs "here's a janky workaround") but such is the current state of things.

[Homebrew also has a concept of versioned formulae for situations where applications have multiple streams of development and support](https://docs.brew.sh/Versions). If the formula meets the criteria to be versioned as such, perhaps that could also be an alternative.

In any case, as you mention, Homebrew is not the only package manager on the block, nor is it well-catered to serve every use case and software delivery paradigm. I encourage you (and all users) to choose tools that work for you and your specific situation.

**Reply by georgalis** [reply]
@alebcay indeed, I read your reply and I was considering my response when the issue was abruptly closed (10 hrs after your response) by the other developer. While you suggested a "reasonable enhancement" to facilitate improvement to the convoluted workflow required to resolve the current vim syntax highlighting issue; that is apparently now off the table, because the feature request is closed, marked done, won't fix.

That reminded me of my last interaction with the developer, in which there was no effort to understand the UX, and like this case, to even consider the user perspective when presented with a scenario. That's fine if everything is golden, however when referenced your comment in a personal email to Mike about closing the issue. He said the problem is not a brew shortcoming, I should write a PR, and don't send a private email to the address he advertised on his github profile.

A review of #6637 and #14809 will reveal the a pattern of disregard for reasonable user expectations. If establishing a rapport around minor issues such as purging orphan/old packages, or even acknowledging that "upstream not evergreen" creates an issue; how can any reasonable resolution be expected for difficult problems? Regardless of the status of the project, that is not a personality I would choose to interact with.

**Reply by alebcay** [reply]
With specific regard to the fact that the issue was closed, usually we only keep issues open if there is a plan (or deemed need) for an immediate fix; the issues board only reflects what's currently receiving the maintainer team's attention. There are times that issues get fixed long after they're closed, or sometimes an issue is opened multiple times before it's deemed large enough to address (and there's someone willing to open a PR for it). As a maintainer team, we generally keep the issues board short to make it easier to focus on what's being worked on.

On the other hand, these discussions here tend to stay available/open and may be more helpful for longer-term discussion that may not be as closely tied to a specific code change, and are an avenue of continuous feedback regardless of whether a feature is considered in-scope or not.

In any case, thank you for your feedback about your experience, including the unpleasant portions.
